### PR TITLE
feat(rotation): sequential drain-first account scheduling mode (#509)

### LINKED ISSUE #509: [feature] Sequential Account Usage Mode (Drain One Account Before Switching)
## Problem Statement

I have two Codex accounts with independent rolling quota windows.

Currently, requests may be distributed across both accounts. While this helps balance usage, it also causes both accounts to consume quota at roughly the same time. As a result, both accounts may enter cooldown or become unavailable around the same period.

For my workflow, a sequential strategy would be more efficient.

Example:

- Start using Account A.
- Continue using Account A until it reaches its quota limit or becomes unavailable.
- Automatically switch to Account B.
- Continue working on Account B while Account A's quota window recovers.

**By the time Account B is heavily utilized, Account A may already have recovered enough quota to become available again.**

This creates a staggered recovery pattern and can provide longer uninterrupted working sessions compared to consuming quota from both accounts simultaneously.

## Proposed Solution

Add an optional scheduling mode such as:

- Sequential
- Drain First

Behavior:

1. Select a primary account.
2. Route all requests to that account while it remains healthy and available.
3. When the account becomes unavailable, delayed, rate limited, or reaches a configurable threshold, automatically switch to the next account.
4. Continue through the account pool sequentially.
5. Once an earlier account recovers, it becomes eligible again.

This mode should be optional and coexist with the current balancing and affinity strategies.

## Alternatives Considered

Current workaround:

- Manually pin Account A.
- Use it until quota is exhausted.
- Manually switch to Account B.
- Repeat throughout the day.

This works well but requires manual intervention.

### In Scope

- Sequential account usage.
- Automatic failover between accounts.
- Optional configurable switch thresholds.
- Compatibility with existing forecasting and health checks.

## Validation Plan

Using two accounts:

1. Enable Sequential / Drain First mode.
2. Generate continuous workload.
3. Verify that all requests initially use Account A.
4. Verify that Account B remains mostly untouched.
5. Verify automatic failover to Account B when Account A becomes unavailable.
6. Verify that Account A can become available again after its quota window recovers.

Expected result:

Instead of both accounts reaching cooldown at roughly the same time, quota recovery becomes staggered, allowing for longer uninterrupted usage.

## Compliance Confirmation

- [x] This proposal is for legitimate personal development workflows.
- [x] This proposal does not request bypassing OpenAI controls or policies.
- [x] This proposal does not introduce multi-tenant resale behavior.

### PR DESCRIPTION
**Pull request by ndycode**
## Summary

Implements the optional Sequential / Drain-First scheduling mode requested in #509.

By default the runtime proxy spreads requests across all available accounts (`hybrid`), so multiple accounts drain at a similar pace and tend to hit cooldown together. This adds an opt-in `schedulingStrategy: "sequential"` that keeps one active account in use until it is fully exhausted, then advances to the next available account. Earlier accounts are reclaimed as soon as their quota window recovers, staggering recovery across the pool for longer uninterrupted sessions.

Resolves the maintainer-confirmed spec from #509:
- Switches only when the active account is 100 percent exhausted (rate-limited / cooling down / circuit-open). No soft early-switch threshold in v1.
- A single active account governs all new requests. Sequential mode intentionally ignores per-session affinity, so once the active account changes every subsequent request follows it regardless of which account started a conversation.
- A manual pin (`codex-multi-auth switch`) always takes precedence; sequential mode is ignored while a pin is active.

## Behavior

- `hybrid` (default, unchanged): weighted health/token/freshness selection across all available accounts.
- `sequential`: sticky on the current active account; scans forward (wrapping the pool) only when the active account is exhausted, and pins the next available account as the new active one.

Default is `hybrid`, so existing deployments are unaffected.

## Changes

- `lib/accounts.ts`: new `getCurrentOrNextForFamilySequential` selector. Sticky on `currentAccountIndexByFamily`; advances only on exhaustion; wraps so a recovered earlier account reclaims the active slot. Per ModelFamily, matching the rest of the selection path.
- `lib/runtime-rotation-proxy.ts`: `chooseAccount` branches on strategy. Pin tier runs first (unchanged); sequential skips the affinity tier; the linear-scan fallback is extracted into a shared helper used by both paths.
- `lib/schemas.ts` / `lib/config.ts`: `schedulingStrategy` enum (`hybrid` | `sequential`, default `hybrid`), `getSchedulingStrategy` accessor following the existing `routingMutex` pattern, `CODEX_AUTH_SCHEDULING_STRATEGY` env override, and a config-explain parity entry.
- `docs/configuration.md`, `docs/reference/settings.md`, `docs/development/CONFIG_FIELDS.md`: documented the setting and env override per the add-config-field runbook.

## Out of scope (v1)

- TUI settings picker: the backend settings hub has no enum-key type yet (existing enum strategies `routingMutex` and `fastSessionStrategy` are also config plus env only). Reachable via settings file and env var.
- Soft early-switch threshold: requester confirmed 100 percent exhaustion as the default. The accessor pattern leaves room to add a threshold later.
- In-flight drain: the proxy is stateless with respect to in-flight request counts, so "drain" means no new selections to the account, not waiting for outstanding requests to finish.

## Testing

- `npm run typecheck`: pass
- `npm run lint:ts`: pass
- `npm test`: 4387 passed, 3 skipped (pre-existing), 0 failures
- `npm run build`: pass

New tests:
- Sequential selector: sticky across calls, advances on exhaustion, wraps to a recovered earlier account, returns null when all exhausted.
- Proxy `chooseAccount`: sequential ignores session affinity, manual pin wins over sequential, advances when the active account is rate-limited.
- Config accessor: default, explicit config, env override both directions, invalid env fallback.
- Updated the five strict default-config `toEqual` assertions in `plugin-config.test.ts` for the new key.

Closes #509

**note:** greptile review for oc-chatgpt-multi-auth. cite files like `lib/foo.ts:123`. confirm regression tests + windows concurrency/token redaction coverage.
---

<h3>Greptile Summary</h3>

adds `schedulingStrategy: "sequential"` (drain-first) as an opt-in alternative to the default `hybrid` account selection mode. the sequential selector sticks to one active account per model family and only advances to the next when the active account is fully exhausted; it wraps the pool so recovered earlier accounts can reclaim the active slot.

- `lib/accounts.ts` – new `getCurrentOrNextForFamilySequential` selector with per-family sticky pointer, policy-block awareness, and wrap-around scan; `lib/runtime-rotation-proxy.ts` – `chooseAccount` branches on strategy, skips session-affinity tier in sequential mode, and `chooseLinearScanFallback` is extracted as a shared helper with an `advanceActivePointer` flag that keeps within-request retries from permanently moving the drain-first primary.
- `lib/schemas.ts` / `lib/config.ts` – `schedulingStrategy` enum field, `getSchedulingStrategy` accessor, and `CODEX_AUTH_SCHEDULING_STRATEGY` env override following the `routingMutex` pattern; three doc files updated.
- two previously-reviewed P1s are addressed: `advanceActivePointer=false` prevents transient retry failures from advancing the sequential pointer, and `blockedIndexes` is now passed into the selector so it never anchors to a policy-blocked account.

<h3>Confidence Score: 5/5</h3>

safe to merge — all three issues caught in the previous review round are fixed, the new selector and proxy branching are correct, and the test suite covers the critical edge cases including the mutex path and transient-failure non-advance regression.

the sequential selector, chooseAccount branching, and routingMutex re-commit skip have all been verified against documented invariants. advanceActivePointer=false correctly prevents within-request retries from permanently moving the drain-first primary, and the blockedIndexes guard prevents the active pointer anchoring on a policy-blocked account. test coverage is thorough across selector unit tests, chooseAccount unit tests, a real-proxy mutex integration test, and config accessor tests. no token safety or windows filesystem risks introduced.

no files require special attention — lib/runtime-rotation-proxy.ts is the most complex change but the advanceActivePointer flag and mutex skip are both well-tested.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| lib/accounts.ts | adds `getCurrentOrNextForFamilySequential` — sticky pointer, wrap-around scan, policy-block guard, and consistent `lastUsed` stamping; `isUsable` conditions align exactly with `getAccountRuntimeSkipReason`, making the outer skip-reason check in chooseAccount a safe belt-and-suspenders guard. |
| lib/runtime-rotation-proxy.ts | sequential branch inserted before session-affinity tier; `chooseLinearScanFallback` extracted with `advanceActivePointer` flag; mutex hot-path skips `markSwitchedLocked` re-commit in sequential mode — all three P1 fixes from earlier review round are present and correct. |
| lib/config.ts | new `getSchedulingStrategy` follows the same `resolveStringSetting` + known-values-set pattern as `getRoutingMutexMode`; env override and invalid-value fallback are both handled; `CONFIG_EXPLAIN_ENTRIES` entry added for parity. |
| lib/schemas.ts | adds `schedulingStrategy: z.enum(["hybrid","sequential"]).optional()` to `PluginConfigSchema`; minimal, correct change. |
| test/accounts.test.ts | 10 new sequential-selector cases covering sticky, exhaustion advance, wrap-around, cooldown, circuit-open, 3-account pool, disabled, unset active index, per-family independence, and the policy-blocked P1 regression; `clearCircuitBreakers()` in `beforeEach` prevents cross-test breaker leakage. |
| test/runtime-rotation-proxy.test.ts | adds sequential proxy tests: affinity-ignored, manual-pin-wins, advances-on-exhaustion, no-pointer-advance-on-transient-failure (P1 regression), all-exhausted; plus the mutex-path integration test driving the real proxy with `routingMutex=enabled` + `schedulingStrategy=sequential`. |
| test/scheduling-strategy-config.test.ts | new file; covers default, explicit config value, env override in both directions, invalid env fallback to config, and invalid persisted config fallback to default — fully matches the routingMutex config-test pattern. |
| test/plugin-config.test.ts | five strict `toEqual` assertions updated to include `schedulingStrategy: "hybrid"` — mechanical but necessary, no issues. |
| docs/configuration.md | added `CODEX_AUTH_SCHEDULING_STRATEGY` to stable env-override table and added the "Sequential / drain-first scheduling" section — correctly placed per the add-config-field runbook. |
| docs/reference/settings.md | added `schedulingStrategy` row to proxy settings table and `CODEX_AUTH_SCHEDULING_STRATEGY` to the env-override list — consistent with existing entries. |
| docs/development/CONFIG_FIELDS.md | added `schedulingStrategy` row to the runtime-only fields table and `CODEX_AUTH_SCHEDULING_STRATEGY` to the env var inventory — correct location and format. |

</details>

<sub>Reviews (4): Last reviewed commit: ["fix(rotation): keep sequential selector ..."](https://github.com/ndycode/codex-multi-auth/commit/459c5ebc762912b699d6cbbc6dc08b86dce3f510) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=35479595)</sub>

### CODE REVIEW

File: lib/runtime-rotation-proxy.ts [hunk 1]
```diff
@@ -1211,6 +1214,44 @@ export function chooseAccount(params: {
...
+		// rate-limited mid-loop): fall through to the shared linear scan below to
+		// find the next eligible account.
+		return chooseLinearScanFallback({
+			accountManager,
+			family,
+			model,
+			attemptedIndexes,
+			policy,
+			skipReasons,
+		});
```
[the remark this answers is not in the document]
Comment (ndycode) [reply lib/runtime-rotation-proxy.ts]: Fixed in 23cfe24. You're right, this was a real regression. `chooseLinearScanFallback` now takes an `advanceActivePointer` flag (default `true` so the hybrid round-robin advance is unchanged), and the sequential path passes `false` so a within-request retry only finds an account to *try* without moving `currentAccountIndexByFamily`. I also guarded the hot-path `markSwitchedLocked` re-commit so it's skipped in sequential mode under `routingMutex=enabled` (the sequential selector already commits the correct active index inside the held mutex).

Added a regression test that reproduces exactly your scenario (active account healthy but in `attemptedIndexes`): it asserts the fallback hands out account 1 to try, the primary stays at 0, and the next fresh request sticks to 0 again. I verified the test fails without the fix (expected 0, got 1) and passes with it.

File: test/runtime-rotation-proxy.test.ts [hunk 2]
```diff
@@ -2406,3 +2408,91 @@ describe("buildTokenInvalidationBody", () => {
...
+			attemptedIndexes: new Set(),
+			now: Date.now(),
+			policy: null,
+			pinnedIndex: null,
+			schedulingStrategy: "sequential",
+		});
+
+		expect(selected?.index).toBe(1);
+	});
+});
```
[the remark this answers is not in the document]
Comment (ndycode) [reply test/runtime-rotation-proxy.test.ts:2622]: Added in 23cfe24. New test `sequential mode survives the routing-mutex path` drives the real proxy via `startRuntimeRotationProxy` with `routingMutex=enabled` and `schedulingStrategy=sequential`, fires three back-to-back requests with distinct session ids, and asserts all three serve `Bearer access-1` and `getActiveIndexForFamily("codex")` stays 0. This exercises the `markSwitchedLocked` re-commit path and confirms it no longer double-advances the sequential primary.

File: lib/accounts.ts [hunk 3]
```diff
@@ -789,6 +789,75 @@ export class AccountManager {
...
+		for (let i = 0; i < count; i++) {
+			const idx = (start + i) % count;
+			const account = this.accounts[idx];
+			if (!isUsable(account)) continue;
+
+			this.currentAccountIndexByFamily[family] = idx;
+			this.cursorByFamily[family] = (idx + 1) % count;
+			account.lastUsed = nowMs();
+			return account;
+		}
```
[the remark this answers is not in the document]
Comment (ndycode) [reply lib/accounts.ts:866]: Intentionally not changing this. The 100-percent-exhaustion behavior is the confirmed spec from #509: the requester explicitly chose to drain each account fully before switching, with no health/forecast early-switch threshold (see their reply on the issue). A soft threshold would defeat the staggered-recovery goal they're after. The `getSchedulingStrategy` accessor and `schedulingDrainThresholdPercent`-style knob were deliberately left out of v1 but the structure leaves room to add an opt-in early-switch threshold later if demand appears. Hybrid mode (the default) keeps the full health-aware selection for everyone who wants it.

File: test/accounts.test.ts [hunk 4]
```diff
@@ -4396,4 +4396,82 @@ describe("AccountManager", () => {
...
+		it("returns null when every account is exhausted", () => {
+			const manager = new AccountManager(undefined, makeStored(2) as never);
+			const account0 = manager.setActiveIndex(0)!;
+			const account1 = manager.getAccountByIndex(1)!;
+			manager.markRateLimited(account0, 60_000, "codex");
+			manager.markRateLimited(account1, 60_000, "codex");
+
+			expect(manager.getCurrentOrNextForFamilySequential("codex")).toBeNull();
+		});
+	});
```
[the remark this answers is not in the document]
Comment (ndycode) [reply test/accounts.test.ts:4628]: Added in 23cfe24. The sequential selector describe block now covers the non-rate-limit exhaustion paths: a cooling-down active account (advance via `markAccountCoolingDown`, then reclaim after `clearAccountCooldown`) and an open-circuit active account (three `recordFailure` calls trip the breaker, selector advances to the next account). Combined with the existing rate-limit and all-exhausted cases, the four `isUsable` rejection reasons are now exercised.

File: test/runtime-rotation-proxy.test.ts [hunk 5]
```diff
@@ -2406,3 +2408,91 @@ describe("buildTokenInvalidationBody", () => {
...
+			model: null,
+			attemptedIndexes: new Set(),
+			now: Date.now(),
+			policy: null,
+			pinnedIndex: null,
+			schedulingStrategy: "sequential",
+		});
+
+		expect(selected?.index).toBe(1);
+	});
```
[the remark this answers is not in the document]
Comment (ndycode) [reply test/runtime-rotation-proxy.test.ts:2557]: Added in 23cfe24. New test `returns null when every account is exhausted in sequential mode` in the `chooseAccount` suite rate-limits both accounts and asserts the call returns null, protecting the activeIndex/affinity/pin flow from regressions.

File: test/scheduling-strategy-config.test.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
+import { describe, it, expect, afterEach, beforeEach } from "vitest";
```
[the remark this answers is not in the document]
Comment (ndycode) [reply test/scheduling-strategy-config.test.ts:1]: Leaving the import as-is. The guideline doesn't match the repo's actual convention: 260 of 275 `*.test.ts` files import `describe`/`it`/`expect` from `vitest` explicitly. Matching the dominant local pattern keeps the file consistent with its neighbors (e.g. `routing-mutex-config.test.ts`, which this file mirrors). `globals: true` is set so the import is redundant, but removing it here alone would make this one file an outlier.

File: test/runtime-rotation-proxy.test.ts [hunk 7]
```diff
@@ -2406,3 +2408,91 @@ describe("buildTokenInvalidationBody", () => {
...
+			activeIndex: 0,
+			activeIndexByFamily: { codex: 0 },
+			accounts: Array.from({ length: count }, (_, i) => ({
+				refreshToken: `token-${i}`,
+				addedAt: now,
+				lastUsed: now - i * 1000,
+			})),
+		};
+		return new AccountManager(undefined, stored as never);
+	};
```
[the remark this answers is not in the document]
Comment (ndycode) [reply test/runtime-rotation-proxy.test.ts:2491]: Keeping the cast for consistency. `as never` for the stored-account fixture is the established convention across `accounts.test.ts` (used in ~30 places there) — these are deliberately minimal V3 fixtures (just `refreshToken`/`addedAt`/`lastUsed`) that omit optional fields the runtime fills in. Switching this one helper to a fully-typed `AccountStorageV3` would diverge from the surrounding suite. Worth a separate cleanup PR across all the fixtures if the team wants typed factories, rather than a one-off here.

File: lib/accounts.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ndycode) [reply lib/accounts.ts:866]: Fixed in 459c5eb. You're right — `isUsable()` gated only on rate-limit / cooldown / circuit / enabled / workspaces, not on `policy.blockedAccountIndexes`, so the selector could commit `currentAccountIndexByFamily` to a paused/drained/capability-blocked account before `chooseAccount` rejected it, permanently anchoring the drain-first primary there.

`getCurrentOrNextForFamilySequential` now takes the request's `blockedIndexes` set and `isUsable()` excludes blocked accounts, so the active pointer never anchors on one. The proxy passes `policy?.blockedAccountIndexes` in. Added a regression test (account 0 exhausted, account 1 policy-blocked → returns null and leaves the pointer at 0; selects account 1 only once the block clears). Verified it fails without the fix (returns account 1 / anchors).

File: test/runtime-rotation-proxy.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ndycode) [reply test/runtime-rotation-proxy.test.ts:2557]: Investigated this closely and it does not reproduce — the precondition is unreachable, so I'm not changing it.

The concern requires the primary to be in `attemptedIndexes` while still `isUsable`. I audited every `continue` branch in the request loop and each one that benches the primary also makes it unavailable in the same step:

- non-retryable / retryable refresh failure (the `!refreshed.retryable` branch you cited): `prepareAccountToken` calls `applyMonotonicAuthCooldown(account, DEFAULT_AUTH_FAILURE_COOLDOWN_MS)` (lib/runtime-rotation-proxy.ts:912) **before** returning, so the account is cooling down.
- 429: `markRateLimitedWithReason(..., "quota")`.
- 5xx / network error: `markAccountCoolingDown(...)`.
- workspace-disabled: `setAccountEnabled(index, false)`.
- 401 invalidation: `applyMonotonicAuthCooldown(...)`.

So after any retry the primary is rate-limited / cooling / disabled, `isUsable()` is false, and the sequential selector legitimately advances to the next account (and reclaims the primary once its window elapses). I verified this empirically: with a non-retryable 400 refresh failure on the primary, the primary ends `isAccountCoolingDown === true` and the advance to the next account is correct.

I had drafted a guard + test for this, then revert-tested and confirmed the "test" only passed because the modeled state (healthy + attempted) can't occur; dropped both to avoid shipping a special-case branch for an unreachable path. Happy to revisit if you can point to a `continue` branch that refunds the primary without also benching it.
